File: detected_faces_viewer.py
import cv2
import pickle
import numpy as np
import matplotlib.pyplot as plt
from matplotlib.widgets import Slider, Button
from matplotlib.patches import Circle
from collections import defaultdict
from pathlib import Path
import random
import math

# Config
PICKLE_PATH = "./pickles/emotion_cache_filtered.pkl"
IMAGE_DIR = "./wikiart/"
GRID_SIZE = 20  # 10x10 emotion grid
EMOTION_AXES = {
    'x': ('angry', 'happy'),  # Left-right axis
    'y': ('sad', 'fear')      # Bottom-top axis
}

def rotate_point(x, y, angle_degrees=45):
    theta = math.radians(angle_degrees)
    cos_t, sin_t = math.cos(theta), math.sin(theta)
    x_new = cos_t * x - sin_t * y
    y_new = sin_t * x + cos_t * y
    return x_new, y_new

class EmotionExplorer:
    def __init__(self):
        self.results = self.load_results()
        self.grid = self.create_emotion_grid()
        self.current_point = (GRID_SIZE//2, GRID_SIZE//2)  # Start at neutral center
        
        # Set up plot
        self.fig = plt.figure(figsize=(16, 10))
        self.ax_img = plt.subplot2grid((3, 3), (0, 0), colspan=2, rowspan=3)
        self.ax_map = plt.subplot2grid((3, 3), (0, 2))
        self.ax_slider_x = plt.subplot2grid((3, 3), (1, 2))
        self.ax_slider_y = plt.subplot2grid((3, 3), (2, 2))
        
        # Add controls
        self.slider_x = Slider(self.ax_slider_x, 'X (Angry ↔ Happy)', 0, GRID_SIZE-1, 
                              valinit=GRID_SIZE//2, valstep=1)
        self.slider_y = Slider(self.ax_slider_y, 'Y (Sad ↔ Fear)', 0, GRID_SIZE-1, 
                              valinit=GRID_SIZE//2, valstep=1)
        self.slider_x.on_changed(self.update_from_slider)
        self.slider_y.on_changed(self.update_from_slider)
        
        # Click interaction
        self.cursor = Circle((0, 0), 0.3, color='red', alpha=0.8)
        self.ax_map.add_patch(self.cursor)
        self.fig.canvas.mpl_connect('button_press_event', self.on_click)
        
        self.setup_emotion_map()
        self.display_random_image()
        
    def load_results(self):
        with open(PICKLE_PATH, "rb") as f:
            return pickle.load(f)
        
    
    def emotion_to_coords(self, emotions):
        """Convert emotion percentages to evenly distributed grid coordinates"""
        # Normalize emotion values to [-1, 1] range
        x_val = (emotions[EMOTION_AXES['x'][1]] - emotions[EMOTION_AXES['x'][0]]) / 100.1
        y_val = (emotions[EMOTION_AXES['y'][1]] - emotions[EMOTION_AXES['y'][0]]) / 100.1
        
        # Map to [0, GRID_SIZE-1] with equal bucket sizes
        # x = int((x_val + 1) * (GRID_SIZE) / 2)
        # y = int((y_val + 1) * (GRID_SIZE) / 2)
        x, y = rotate_point(x_val, y_val, 45)  # Rotate the point by 45 degrees
        x = int((x + 1) * (GRID_SIZE) / 2)
        y = int((y + 1) * (GRID_SIZE) / 2)
        
        # Ensure we stay within grid bounds
        return (
            np.clip(x, 0, GRID_SIZE - 1),
            np.clip(y, 0, GRID_SIZE - 1)
        )
    
    def create_emotion_grid(self):
        """Bucket all images into emotion grid"""
        grid = defaultdict(list)
        for img_path, data in self.results.items():
            if data:
                x, y = self.emotion_to_coords(data['emotion'])
                grid[(x, y)].append(img_path)
        return grid
    
    def setup_emotion_map(self):
        """Create heatmap of available emotions"""
        heatmap = np.zeros((GRID_SIZE, GRID_SIZE))
        for (x, y), imgs in self.grid.items():
            heatmap[y, x] = math.log(len(imgs) + 1)  # Log scale for visibility
        
        self.ax_map.imshow(heatmap, cmap='viridis', extent=[0, GRID_SIZE, 0, GRID_SIZE],
                          origin='lower')  # Fix origin to match sliders
        self.ax_map.set_title(f"Click anywhere!\n{EMOTION_AXES['x'][0]} ↔ {EMOTION_AXES['x'][1]}\n{EMOTION_AXES['y'][0]} ↔ {EMOTION_AXES['y'][1]}")
        self.ax_map.grid(alpha=0.3)
    
    def update_selection(self, x, y):
        """Update current selection point"""
        self.current_point = (x, y)
        self.cursor.center = (x + 0.5, y + 0.5)  # Center in grid cell
        self.display_random_image()
        self.fig.canvas.draw_idle()
    
    def update_from_slider(self, val):
        """Handle slider changes"""
        self.update_selection(int(self.slider_x.val), int(self.slider_y.val))
    
    def on_click(self, event):
        """Handle map clicks"""
        if event.inaxes == self.ax_map:
            x = int(event.xdata)
            y = int(event.ydata)
            self.slider_x.set_val(x)
            self.slider_y.set_val(y)
            self.update_selection(x, y)
            # print(f"Clicked at: {event.xdata:.2f}, {event.ydata:.2f}")
    
    def display_random_image(self):
        """Show random image from current emotion bucket"""
        self.ax_img.clear()
        
        x, y = self.current_point
        available_imgs = self.grid.get((x, y), [])
        
        if not available_imgs:
            self.ax_img.text(0.5, 0.5, "No images in this emotion zone", 
                           ha='center', va='center', fontsize=12)
            self.ax_img.set_title(f"Zone {x},{y}: No images")
            return
        
        img_path = random.choice(available_imgs)
        data = self.results[img_path]
        img = cv2.cvtColor(cv2.imread(img_path), cv2.COLOR_BGR2RGB)
        
        # Draw face box
        face = data['face_region']
        cv2.rectangle(img, (face['x'], face['y']), 
                     (face['x']+face['w'], face['y']+face['h']), 
                     (0, 255, 0), 2)
        
        self.ax_img.imshow(img)
        self.ax_img.set_title(
            f"Zone {x},{y}: {len(available_imgs)} images\n"
            f"Dominant: {data['dominant']} ({data['emotion'][data['dominant']]:.1f}%)"
        )
        self.ax_img.axis('off')
        
        # Print emotion stats
        print(f"\n📍 Selected zone: {x},{y}")
        print(f"📊 Images in zone: {len(available_imgs)}")
        print("🎭 Emotion scores:")
        for e, score in sorted(data['emotion'].items(), key=lambda x: -x[1]):
            print(f"  - {e}: {score:.1f}%")

if __name__ == "__main__":
    explorer = EmotionExplorer()
    plt.tight_layout()
    plt.show()

File: detected_faces_viewer.js
// Config
const GRID_SIZE = 20;
const EMOTION_AXES = {
  x: ['angry', 'happy'],
  y: ['sad', 'fear']
};

// List your JSON files here (adjust paths to your JSON files)
const JSON_FILES = [
  'emotion_json/Abstract_Expressionism_emotion_cache.json',
  'emotion_json/Abstract_Expressionism_emotion_cache_filtered.json',
  // Add more as needed
];

let results = {};
let gridBuckets = new Map();
let selectedCell = null;

// Rotate point by 45 degrees
function rotatePoint(x, y, angleDegrees = 45) {
  const rad = (angleDegrees * Math.PI) / 180;
  const cosT = Math.cos(rad);
  const sinT = Math.sin(rad);
  return {
    x: cosT * x - sinT * y,
    y: sinT * x + cosT * y,
  };
}

// Map emotion to grid coords
function emotionToCoords(emotion) {
  const xVal = (emotion[EMOTION_AXES.x[1]] - emotion[EMOTION_AXES.x[0]]) / 100.1;
  const yVal = (emotion[EMOTION_AXES.y[1]] - emotion[EMOTION_AXES.y[0]]) / 100.1;
  const rotated = rotatePoint(xVal, yVal, 45);
  const x = Math.min(GRID_SIZE - 1, Math.max(0, Math.floor(((rotated.x + 1) * GRID_SIZE) / 2)));
  const y = Math.min(GRID_SIZE - 1, Math.max(0, Math.floor(((rotated.y + 1) * GRID_SIZE) / 2)));
  return { x, y };
}

// Load JSON file
async function loadJson(url) {
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`Failed to load JSON: ${url}`);
  return resp.json();
}

// Merge multiple JSON files into one object
async function loadAllJsonFiles(urls) {
  const combined = {};
  for (const url of urls) {
    try {
      const data = await loadJson(url);
      Object.assign(combined, data); // merge all key-value pairs
    } catch (err) {
      console.warn(err.message);
    }
  }
  return combined;
}

// Build bucket map (x,y) => [imgPaths]
function buildGridBuckets(data) {
  const buckets = new Map();
  for (const [imgPath, det] of Object.entries(data)) {
    if (!det || !det.emotion) continue;
    const { x, y } = emotionToCoords(det.emotion);
    const key = `${x},${y}`;
    if (!buckets.has(key)) buckets.set(key, []);
    buckets.get(key).push(imgPath);
  }
  return buckets;
}

// Render grid heatmap
function renderGrid() {
  const gridDiv = document.getElementById('grid');
  gridDiv.innerHTML = '';
  for (let y = 0; y < GRID_SIZE; y++) {
    for (let x = 0; x < GRID_SIZE; x++) {
      const key = `${x},${y}`;
      const count = gridBuckets.get(key)?.length || 0;
      const cell = document.createElement('div');
      cell.classList.add('cell');
      const brightness = count > 0 ? Math.min(1, Math.log(count + 1) / 3) : 0;
      const green = Math.floor(255 * brightness);
      cell.style.backgroundColor = `rgb(0, ${green}, 0)`;
      cell.title = `Cell ${x},${y}: ${count} images`;
      cell.dataset.key = key;
      cell.onclick = () => selectCell(x, y);
      gridDiv.appendChild(cell);
    }
  }
}

// Select a cell, highlight it, show image + face box
function selectCell(x, y) {
  if (selectedCell) {
    const prev = document.querySelector(`.cell.selected`);
    if (prev) prev.classList.remove('selected');
  }
  const key = `${x},${y}`;
  selectedCell = key;
  const cell = document.querySelector(`.cell[data-key='${key}']`);
  if (cell) cell.classList.add('selected');
  document.getElementById('selected-cell').textContent = key;
  showRandomImage(key);
}

// Show a random image for selected bucket with face box overlay
function showRandomImage(key) {
  const paths = gridBuckets.get(key);
  const faceImg = document.getElementById('face-image');
  const canvas = document.getElementById('face-canvas');
  const ctx = canvas.getContext('2d');
  const infoDiv = document.getElementById('emotion-details');

  if (!paths || paths.length === 0) {
    faceImg.src = '';
    infoDiv.textContent = 'No images in this cell.';
    ctx.clearRect(0, 0, canvas.width, canvas.height);
    return;
  }

  const imgPath = paths[Math.floor(Math.random() * paths.length)];
  const data = results[imgPath];

  faceImg.onload = () => {
    // Match canvas size exactly to loaded image size
    canvas.width = faceImg.width;
    canvas.height = faceImg.height;

    ctx.clearRect(0, 0, canvas.width, canvas.height);

    if (data.face_region) {
      const fr = data.face_region;
      ctx.lineWidth = 3;
      ctx.strokeStyle = 'lime';
      ctx.strokeRect(fr.x, fr.y, fr.w, fr.h);
    }
  };

  faceImg.onerror = () => {
    infoDiv.textContent = 'Failed to load image.';
    ctx.clearRect(0, 0, canvas.width, canvas.height);
  };

  faceImg.src = imgPath;

  infoDiv.innerHTML =
    `<strong>Dominant:</strong> ${data.dominant}<br/>Emotions:<br/>` +
    Object.entries(data.emotion)
      .sort((a, b) => b[1] - a[1])
      .map(([e, v]) => `${e}: ${v.toFixed(1)}%`)
      .join('<br/>');
}

// Initialization
(async () => {
  try {
    results = await loadAllJsonFiles(JSON_FILES);
    gridBuckets = buildGridBuckets(results);
    renderGrid();
  } catch (e) {
    alert('Error loading data: ' + e.message);
  }
})();
